fix: Legend does not overflow for categorical legend type

vidhi04bhatia opened the pull request:
**Related Ticket:** _#1174_

### Description of Changes
1. Added logic to show chevron arrows only for legend `type: categorical`.
2. Modified logic for labels `LegendList` and `LegendSwatch` to handle for both legend types
3. Added a new interface `LegendListProps` to handle the above mentioned point 2
4. Changed logic in `LayerCategoricalGraphic` to display the swatch as a list 

### Notes & Questions About Changes
Do we show the scrollbar so that one knows there are more data? Or keep it transparent?

### Validation / Testing
[Data catalog preview link with categorical colormap](https://deploy-preview-1610--veda-ui.netlify.app/exploration?search=&datasets=%5B%7B%22id%22%3A%22no2-monthly-diff%22%2C%22settings%22%3A%7B%22isVisible%22%3Atrue%2C%22opacity%22%3A100%2C%22analysisMetrics%22%3A%5B%7B%22id%22%3A%22mean%22%2C%22label%22%3A%22Average%22%2C%22chartLabel%22%3A%22Average%22%2C%22themeColor%22%3A%22infographicB%22%7D%2C%7B%22id%22%3A%22std%22%2C%22label%22%3A%22St+Deviation%22%2C%22chartLabel%22%3A%22St+Deviation%22%2C%22themeColor%22%3A%22infographicD%22%7D%5D%2C%22colorMap%22%3A%22rdbu_r%22%7D%7D%5D&taxonomy=%7B%7D&date=2023-12-01T06%3A00%3A00.000Z&dateRange=)
[Story preview link with categorical colormap](https://deploy-preview-1610--veda-ui.netlify.app/stories/air-quality-and-covid-19) (shows how it looks in `Scrollytelling` and `Map` components)

In this story test for:
1. Legend should be expanded by default and displaying the swatch along with the label corresponding to it.
<img width="1728" alt="Screenshot 2025-04-11 at 7 48 37 PM" src="https://github.com/user-attachments/assets/9ae3f4a0-b44a-4751-9117-7827a80a2447" />

2. On clicking the Chevron button, the legend should collapse.
<img width="1726" alt="Screenshot 2025-04-11 at 7 48 49 PM" src="https://github.com/user-attachments/assets/a501c900-1f68-4301-b201-907147eea695" />

3. The Chevron behavior should not affect the Info Icon.
4. Legend container should be vertically scrollable

slesaad commented:
1. It looks like in the collapsed view, you don't see any color. I think it's valuable to have the colors in the collapsed view as well. And when you hover on the color, show the relevant category as a tooltip if available.  One concern there is having too many colors to fit, but you can always squeeze it in based on the number of colors.
2. The legend also appears on the exploration page, and there's no option to collapse/expand it there. I also think the default should be collapsed in the exploration page.
![image](https://github.com/user-attachments/assets/3ccde22c-6c69-45e4-8c19-050a0e7e23bc)

vidhi04bhatia commented:
Thank you for your feedback. I will look into it and resolve those.

vidhi04bhatia commented:
1. The color is now visible without overflowing. I added a new tag and property SwatchSegment. This is how it looks like now;
<img width="1707" alt="Screenshot 2025-04-16 at 4 41 35 PM" src="https://github.com/user-attachments/assets/4dec3331-405d-4b52-8667-6b446e656b45" />

<img width="1710" alt="Screenshot 2025-04-16 at 4 41 25 PM" src="https://github.com/user-attachments/assets/36bd0e91-41a3-4e5b-b04e-007a2b2ffd7e" />
2. The color and chevron arrows has been added to the exploration page. I think we should not have the legend items collapse. So the user knows we have added a list for their convenience, If they do not want to see the list, they could click on the chevron arrow and collapse it. But the color segment would still be visible. On hovering over the colors, they could see the labels corresponding to that color. 
<img width="1707" alt="Screenshot 2025-04-16 at 4 40 21 PM" src="https://github.com/user-attachments/assets/ddb7683c-b3bc-4778-a61e-7040a46646b2" />
3. The question marks issue has been fixed.

slesaad commented:
This is awesome. thanks @vidhi04bhatia ! 
Just requesting some small changes:
1. 
![image](https://github.com/user-attachments/assets/fe68b33c-0b11-497c-a07f-33517ade4858)
let's vertically center align the arrow with the info icon. see how it's aligned at the top, right now?
2. The default expand behavior for the mdx component (in stories) makes sense, but let's have the default be collapsed in the data catalog page.
3. The direction of the chevron arrow in the data catalog page needs to be reversed. 

@faustoperez would love it if you could review the design. this is based on your recommendation in [this thread](https://github.com/NASA-IMPACT/veda-ui/issues/1174#issuecomment-2431329919).

vidhi04bhatia commented:
Thank you for the feedback!
I have fixed the issues:
1. the arrow now aligns with the icon 
<img width="52" alt="Screenshot 2025-04-17 at 10 48 32 AM" src="https://github.com/user-attachments/assets/66ccefe8-bf73-4eae-a059-c277774b6746" />
2. changed the default behavior on data catalog for it to collapse the legend list.
3. the direction of the chevron icon on data catalog has been reversed.

slesaad reviewed (approved):
Looks good to me!

AliceR reviewed (approved):
Hi! Thanks a lot for working on improving the legend — this has been a longstanding issue, and it’s great to see some movement on it. The visual result looks solid (and already has some approvals), so I’ll leave the UI side of things to others to weigh in on.

From a code perspective: you’ve made it work, which is great! That said, there are quite a few conditionals and ternary operators scattered throughout the legend-related files. One way to clean this up could be by separating the different types of legends into dedicated components.

For example, you could define a base styled component like `const BaseLegendList = styled.dl...`, and then build on that with `const CategoricalLegend = styled(BaseLegendList)...` and `const GradientLegend = styled(BaseLegendList)...`. The same pattern could be applied to other elements like `LegendSwatches`.

This way, the logic for choosing which type of legend to render is handled by a single conditional at a higher level, keeping the internals of each component simpler and more focused. It should make the code easier to maintain and reason about down the line.

That said, I’ll go ahead and approve this — the UI improvement is significant, and this already brings things forward. But I do think it would be worth revisiting the structure for maintainability in a follow-up. Let’s keep that in mind as we iterate!

slesaad commented:
Thanks, @AliceR, for the review. I believe the reiteration will also come up when moving the legend component to USWDS, so I agree that we should go ahead and merge this as is for now.
